mdp/nodes/sfa_nodes.py

```diff
@@ -134,22 +169,25 @@ def _train(self, x, include_last_sample=None):
 
     def _stop_training(self, debug=False):
         ##### request the covariance matrices and clean up
-        self.cov_mtx, self.avg, self.tlen = self._cov_mtx.fix()
-        del self._cov_mtx
+        if hasattr(self, '_dcov_mtx'):
+            self.cov_mtx, self.avg, self.tlen = self._cov_mtx.fix()
+            del self._cov_mtx
```

Comment:
What necessitates the `hasattr()` call here..?

Reply:
With that check it is possible to rerun a failed stop_training call with a different `_sfa_solver`, e.g. one of those supporting `rank_deficit`. Without the check, the whole training phase would need to be repeated.

Reply:
Ah, I see. Hm. Is that what we want? This would modify the training procedure on the fly, without giving the user control over this. My preference would be to let it crash, and allow the user to adapt the whole procedure until a configuration is found that runs stable for their respective use case. Not sure I have a strong opinion about this though, it only seems like it might take away an option that I'd like to have control over myself -- but then again, that behavior can of course be optional.. which, however, would also make the whole node more complicated.

Reply:
after a failed stop_training the node is usually garbage, isn't it? The user can still decide whether to redo the training phase on a fresh node or whether to recycle the garbage by rerunning stop_training.

Reply:
It might be useful to create backups after each successful training phase and store them in some temp folder. Say two training phases are successful, leaving behind two backup files, and the third phase crashes. The error message would tell the user where to find the backups and training can resume with different parameters for the problematic layer. That would be the same functionality (if I understand it correctly) but leave the control over what happens / how training is executed to the user.

Reply:
There is a system already in place to do this kind of things: look for `CheckpointFlow` and `Flow.set_crash_recovery` (all of this documented in the Flow class docstring). This is done at the flow level, because there is where you typically do training and testing nodes. Additionally, I don't understand what you mean that after a failed stop_training the node is garbage. If you used `stop_training(debug=True)` all internal attributes are stored, so you can actually re-run `stop_training` after having changed something else in the node, for example `node._symeig`(or `_eig_solver` or whatever you want to call it)

Reply:
You would need to manually restore the `_cov_mtx` and `_dcov_mtx` attributes from `cov_mtx` and `dcov_mtx`, even if running with `debug=True`, because `del self._cov_mtx` and `del self._dcov_mtx` are unconditionally executed before we know whether `stop_training()` succeeds. On another call it would crash on trying to execute `self.cov_mtx, self.avg, self.tlen = self._cov_mtx.fix()`.
While possible, this kind of manual fix requires knowledge about the node's internals and a user couldn't figure it out without looking into the code. (Things like this are the reason why I have to look into MDP code frequently, which isn't a good user experience.)

I find the backup idea and `CheckpointFlow` stuff much more advanced and complicated than adding these simple existence checks for `_cov_mtx` and `_dcov_mtx`. Really don't see why they would hurt so much or why they would reduce user control.

Reply:
Sorry, there is a typo in https://github.com/Stewori/mdp-toolkit/blob/e887365a00efe14307099e159b82231407fc7028/mdp/nodes/sfa_nodes.py#L172
Of course it should be
`if hasattr(self, '_cov_mtx'):` rather than
`if hasattr(self, '_dcov_mtx'):`.

Reply:
> Things like this are the reason why I have to look into MDP code frequently, which isn't a good user experience.

Same here (and in our office in general). I feel like this should be a fundamental guideline for any upcoming changes (in both the code and the documentation), to help alleviate this as much as we can.

As for a system to redo training phases on the fly: I'm in two minds about the whole idea; too much of such "here, let me help you"-features can quickly take away important decisions of the design process. But as long as it stays optional it shouldn't hurt either, and the `CheckpointFlow` seems to completely outsource it, too.